routing/bicycle_directions.cpp

```diff
@@ -222,39 +165,172 @@ Index::FeaturesLoaderGuard & BicycleDirectionsEngine::GetLoader(MwmSet::MwmId co
+  return true;
+}
+
+void BicycleDirectionsEngine::FillPathSegmentsAndAdjacentEdgesMap(
+    RoadGraphBase const & graph, vector<Junction> const & path,
+    IRoadGraph::TEdgeVector const & routeEdges, my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_GREATER(pathSize, 1, ());
+  // Filling |m_adjacentEdges|.
+  m_adjacentEdges.insert(make_pair(UniNodeId(UniNodeId::Type::Mwm), AdjacentEdges(0)));
+  uint32_t constexpr kInvalidSegId = std::numeric_limits<uint32_t>::max();
+  // |startSegId| is a value to keep start segment id of a new instance of LoadedPathSegment.
+  uint32_t startSegId = kInvalidSegId;
+  vector<Junction> prevJunctions;
+  vector<Segment> prevSegments;
+  for (size_t i = 1; i < pathSize; ++i)
+  {
+    if (cancellable.IsCancelled())
+      return;
+
+    Junction const & prevJunction = path[i - 1];
+    Junction const & currJunction = path[i];
+    IRoadGraph::TEdgeVector ingoingEdges;
+    IRoadGraph::TEdgeVector outgoingEdges;
+    graph.GetOutgoingEdges(currJunction, outgoingEdges);
+    graph.GetIngoingEdges(currJunction, ingoingEdges);
+    ASSERT_EQUAL(routeEdges.size(), pathSize - 1, ());
+
+    Edge const & inEdge = routeEdges[i - 1];
+    // Note. |inFeatureId| may be invalid in case of adding fake features.
+    // It happens for example near starts and a finishes.
+    FeatureID const & inFeatureId = inEdge.GetFeatureId();
+    uint32_t const inSegId = inEdge.GetSegId();
+    bool const inIsForward = inEdge.IsForward();
+
+    if (startSegId == kInvalidSegId)
+      startSegId = inSegId;
+
+    if (inFeatureId.IsValid() && i + 1 != pathSize && !IsJoint(ingoingEdges, outgoingEdges, inEdge, routeEdges[i]))
+    {
+      prevJunctions.push_back(prevJunction);
+      Segment inSegment;
+      if (GetSegment(inFeatureId, inSegId, inIsForward, inSegment))
+        prevSegments.push_back(inSegment);
+      continue;
+    }
+    CHECK_EQUAL(prevJunctions.size(), abs(static_cast<int32_t>(inSegId - startSegId)), ());
+
+    prevJunctions.push_back(prevJunction);
+    prevJunctions.push_back(currJunction);
+    Segment segment;
+    if (GetSegment(inFeatureId, inSegId, inIsForward, segment))
+      prevSegments.push_back(segment);
+
+    AdjacentEdges adjacentEdges(ingoingEdges.size());
+    UniNodeId uniNodeId(UniNodeId::Type::Mwm);
+    GetUniNodeIdAndAdjacentEdges(outgoingEdges, inFeatureId, startSegId, inSegId + 1, inIsForward,
+                                 uniNodeId, adjacentEdges);
+
+    size_t const prevJunctionSize = prevJunctions.size();
+    LoadedPathSegment pathSegment(UniNodeId::Type::Mwm);
+    LoadPathAttributes(uniNodeId.GetFeature(), pathSegment);
+    pathSegment.m_nodeId = uniNodeId;
+    pathSegment.m_path = std::move(prevJunctions);
+    // @TODO(bykoianko) |pathSegment.m_weight| should be filled here.
+
+    // Checking if |prevJunctions| reflects real feature ids. If not it means that |prevJunctions|
```

Comment:
I'm putting the following comment instead:
```
    // For LoadedPathSegment instances which contain fake feature id(s), |LoadedPathSegment::m_trafficSegs|
    // should be empty because there's no traffic information for fake features.
    // Traffic information is not supported for bicycle routes as well.
    // Method BicycleDirectionsEngine::GetSegment() returns false for fake features and for bicycle routes.
    // It leads to preventing pushing item to |prevSegments|. So if there's no enough items in |prevSegments|
    // |pathSegment.m_trafficSegs| should be empty.
    // Note. For the time being BicycleDirectionsEngine is used for turn generation for bicycle and car routes.
```